src/sinks/util/buffer/metrics.rs

```diff
@@ -281,6 +182,134 @@ impl Batch for MetricBuffer {
+    }
+
+    /// Either pass the metric through as-is if absolute, or convert it
+    /// to absolute if incremental.
+    pub fn make_absolute(&mut self, metric: Metric) -> Option<Metric> {
+        match metric.data.kind {
+            MetricKind::Absolute => Some(metric),
+            MetricKind::Incremental => self.incremental_to_absolute(metric),
+        }
+    }
+
+    /// Either convert the metric to incremental if absolute, or
+    /// aggregate it with any previous value if already incremental.
+    pub fn make_incremental(&mut self, metric: Metric) -> Option<Metric> {
+        match metric.data.kind {
+            MetricKind::Absolute => self.absolute_to_incremental(metric),
+            MetricKind::Incremental => self.aggregate_incremental(metric),
+        }
+    }
+
+    /// Convert the incremental metric into an absolute one, using the
+    /// state buffer to keep track of the value throughout the entire
+    /// application uptime.
+    fn incremental_to_absolute(&mut self, metric: Metric) -> Option<Metric> {
+        let mut entry = MetricEntry(metric.into_absolute());
+        let mut existing = self.0.take(&entry).unwrap_or_else(|| {
+            // Start from zero value if the entry is not found.
+            MetricEntry(entry.zero())
+        });
+        existing.data.value.add(&entry.data.value);
+        entry.data.value = existing.data.value.clone();
+        self.0.insert(existing);
+        Some(entry.0)
+    }
+
+    /// Convert the absolute metric into an incremental by calculating
+    /// the increment from the last saved absolute state.
+    fn absolute_to_incremental(&mut self, metric: Metric) -> Option<Metric> {
+        let mut entry = MetricEntry(metric);
+        match self.0.get(&entry) {
+            Some(reference) => {
+                // From the stored reference value, emit an increment
+                entry.data.value.subtract(&reference.data.value);
+                Some(entry.0.into_incremental())
+            }
+            None => {
+                // No reference so store this and emit nothing
+                self.0.insert(entry);
+                None
+            }
+        }
+    }
+
+    /// Add the saved state to this incremental metric, saving it as a
+    /// starting point if no previous state is present.
+    pub fn aggregate_incremental(&mut self, metric: Metric) -> Option<Metric> {
+        let mut entry = MetricEntry(metric);
+        let new_entry = match self.0.take(&entry) {
+            Some(mut previous) => {
+                previous.data.add(&entry.data);
+                entry.data = previous.data.clone();
+                previous
+            }
+            None => entry.clone(),
+        };
+        self.0.replace(new_entry);
+        Some(entry.0)
+    }
+
+    /// Produce a fresh state by cloning all the absolute references and
+    /// updating them with any increment previously emitted in this batch.
+    fn fresh(&self, metrics: &Self) -> Self {
```

Comment:
I'm a little bit confused as to what this method is doing, although I could be misunderstanding something. As far as I can tell:

The only user of MetricSet is MetricsBuffer which on receiving a metric, feeds this into MetricNormalize::apply_state with a state MetricSet. Any result it will store in a separate `self.metrics` replacing any prior contents. There are four options depending on what it is receiving and what it outputs

**1. Absolute to Absolute**

Doesn't touch state, propagated directly into self.metrics

**2. Absolute to Incremental**

The first Absolute metric is recorded in the state, with future increments computed from this by `absolute_to_incremental` and stored in self.metrics

**3. Incremental to Incremental**

Accumulates incremental metrics within state with `aggregate_incremental`, storing the accumulated total in self.metrics

**4. Incremental to Absolute**

Accumulates absolute value within state, storing the accumulated total in self.metrics

I think the purpose is for the next buffer as created by `fresh` to handle case 2. correctly. Case 1. doesn't touch the state and case 3. is excluded by the filter. Case 4, however, I think is incorrect because entry.data.value.add doesn't check that the metric being added is actually incremental - `metric.data.add` would work I think (this seems like a fairly fun foot-gun that I've actually removed in the branch in which I'm working on #6045)

It is also perhaps worth noting that whilst BatchSink calls fresh via fresh_replace, PartitionBatchSink, as used by prometheus_remote_write, does not and just calls finish and drops the buffer. I think this will break this component.

Reply:
The purpose of this method is to handle all four cases.

I am not seeing the issue you are describing with case 4. It is true that `MetricValue::add` does not check the metric kind, as that is a separate field, but hashing and comparison on the `MetricEntry` does apply that field, which is why I do that dance with `.into_incremental()` and `.into_absolute()`.

So, when converting an incremental metric to absolute, `incremental_to_absolute` stores the current accumulated value of the incremental metric in the state set, and emits an absolute metric into the metrics set with the same value. When a fresh state set is produced, those accumulated absolute values are cloned, but the update step does not do anything with the emitted absolute metrics, as the `metrics.get` call will not return them.

Your observation about `PartitionBatchSink` is a good insight, and likely explains the problem we saw in #5916 caused by `BatchSink` being replaced by `PartitionBatchSink`. As that is somewhat unrelated to this work, I will open up a new issue for it.

Reply:
Ref: https://github.com/timberio/vector/issues/6062#issuecomment-771009689

Reply:
>  but hashing and comparison on the MetricEntry does apply that field, 

You are quite right, I missed this subtlety of the implementation. As you say, whilst the absolute value stored by 4. will get past the filter it will not find a corresponding incremental metric within `metrics`. :+1: